engine/src/main/java/org/terasology/engine/modes/loadProcesses/CreateWorldEntity.java

```diff
@@ -52,64 +55,101 @@ public String getMessage() {
 
     @Override
     public boolean step() {
+        if (worldEntityExists()) {
+            EntityRef worldEntity = getExistingWorldEntity();
+            configureChunkProvider(worldEntity);
+            useConfigurationOfCurrentWorld(worldEntity);
+        } else {
+            EntityRef worldEntity = createWorldPoolsAndEntity();
+            configureChunkProvider(worldEntity);
+            WorldConfigurator worldConfigurator = useConfigurationFromConfig();
+            configurateWorldEntity(worldConfigurator, worldEntity);
+        }
+
+        return true;
+    }
+
+    private boolean worldEntityExists() {
         EntityManager entityManager = context.get(EntityManager.class);
-        ChunkProvider chunkProvider = context.get(ChunkProvider.class);
+        Iterable<EntityRef> worldEntityIterator = entityManager.getEntitiesWith(WorldComponent.class);
+        return worldEntityIterator.iterator().hasNext();
+    }
 
-        Iterator<EntityRef> worldEntityIterator = entityManager.getEntitiesWith(WorldComponent.class).iterator();
-        if (worldEntityIterator.hasNext()) {
-            EntityRef worldEntity = worldEntityIterator.next();
-            chunkProvider.setWorldEntity(worldEntity);
-
-            // get the world generator config from the world entity
-            // replace the world generator values from the components in the world entity
-            WorldGenerator worldGenerator = context.get(WorldGenerator.class);
-            WorldConfigurator worldConfigurator = worldGenerator.getConfigurator();
-            Map<String, Component> params = worldConfigurator.getProperties();
-            for (Map.Entry<String, Component> entry : params.entrySet()) {
-                Class<? extends Component> clazz = entry.getValue().getClass();
-                Component comp = worldEntity.getComponent(clazz);
-                if (comp != null) {
-                    worldConfigurator.setProperty(entry.getKey(), comp);
-                }
+    private void useConfigurationOfCurrentWorld(EntityRef worldEntity) {
+        // replace the world generator values from the components in the world entity
+        WorldConfiguratorConfigurator configuratorConfigurator = new WorldConfiguratorConfigurator() {
+            @Override
+            public <T extends Component> T getComponentOfWorldEntity(String key, Class<T> clazz) {
+                return worldEntity.getComponent(clazz);
```

Comment:
I tried to use a lambda but it said something about target parameter being generic?

would love it if this didn't take five lines or so. I miss Kotlin.

Reply:
by making the signature less precise and saying it returns `Component` instead of specifically `T`, java is now willing to let me use a lambda here.